src/Pos/Communication/Relay.hs

```diff
@@ -122,104 +130,214 @@ handleReqL
 handleDataL
-    :: ( Bi (DataMsg contents)
-       , Bi (InvMsg key tag)
-       , MonadDHT m
-       , Relay m tag key contents
-       , WorkMode ssc m
-       -- [CSL-659] remove after rewriting to conv
-       , Bi NOP
-       , Message NOP
-       )
+      :: forall tag key contents ssc m .
+      ( Bi (InvMsg key tag)
+      , Bi (ReqMsg key tag)
+      , Bi (DataMsg contents)
+      , MonadDHT m
+      , MessagePart tag
+      , MessagePart contents
+      , Relay m tag key contents
+      , WorkMode ssc m
+      -- [CSL-659] remove after rewriting to conv
+      , Bi NOP
+      , Message NOP
+      )
     => RelayProxy key tag contents
-    -> (ListenerSpec m, OutSpecs)
-handleDataL proxy = listenerOneMsg outSpecs $
-  \_ _ sendActions msg@(DataMsg {..}) -> do
-      key <- contentsToKey dmContents
-      let _ = dataCatchType proxy msg
-          handleDataLDo = do
-              shouldPropagate <- ncPropagation <$> getNodeContext
-              if shouldPropagate then do
-                  logInfo $ sformat
-                      ("Adopted data "%build%" "%
-                       "for key "%build%", propagating...")
-                      dmContents key
-                  tag <- contentsToTag dmContents
-                  sendToNeighbors sendActions $ InvMsg tag (one key)
-              else do
-                  logInfo $ sformat
-                      ("Adopted data "%build%" for "%
-                       "key "%build%", no propagation")
-                      dmContents key
-      processMessage "Data" dmContents verifyDataContents $
-          ifM (handleData dmContents)
-              handleDataLDo $
-              logDebug $ sformat
-                  ("Ignoring data "%build%" for key "%build)
-                  dmContents key
+    -> DataMsg contents
+    -> m ()
+handleDataL proxy msg@(DataMsg {..}) =
+    processMessage () "Data" dmContents verifyDataContents $ do
+        let _ = dataCatchType proxy msg
+        dmKey <- contentsToKey dmContents
+        ifM (handleData dmContents)
+            (handleDataLDo dmKey) $
+                logDebug $ sformat
+                    ("Ignoring data "%build%" for key "%build) dmContents dmKey
   where
-    outSpecs = toOutSpecs [ oneMsgH (invMsgProxy proxy) ]
-
-data RelayProxy key tag contents = RelayProxy
-
-invCatchType :: RelayProxy key tag contents -> InvMsg key tag -> ()
-invCatchType _ _ = ()
-invMsgProxy :: RelayProxy key tag contents -> Proxy (InvMsg key tag)
-invMsgProxy _ = Proxy
-
-reqCatchType :: RelayProxy key tag contents -> ReqMsg key tag -> ()
-reqCatchType _ _ = ()
-reqMsgProxy :: RelayProxy key tag contents -> Proxy (ReqMsg key tag)
-reqMsgProxy _ = Proxy
+    handleDataLDo dmKey = do
+        shouldPropagate <- ncPropagation <$> getNodeContext
+        if shouldPropagate then do
+            tag <- contentsToTag dmContents
+            let inv :: InvOrData tag key contents
+                inv = Left $ InvMsg tag (one dmKey)
```

Comment:
I don't get it? Why put `Left` to queue and explicitly fail on receiving `Right` in worker?